### PR TITLE
Variable declaration

### PR DESCRIPTION
**Pull request by 333fred**
Adds support for IVariableDeclarationGroup, IVariableDeclaration, and IVariableDeclarator to control flow graph generation.

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIfStatement.cs
Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIfStatement.cs]: I recall us talking to rewrite variable initialization into simple assignment, but I am not sure why are we removing the IVariableDeclaration operations itself? I am not sure if this aids any further in control flow analysis. In fact, if some client wants to do some variable analysis and report a diagnostic on the variable declaration, removing the `IVariableDeclarationOperation` nodes makes it difficult for them.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIfStatement.cs
Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIfStatement.cs]: Why are we retaining the `IVariableInitializerOperation`, which is normally expected to be parented by a variable declaration? I thought we were going to do the reverse lowering - remove `IVariableInitializerOperation` nodes and add subsequent simple assignments operations for initialization.

Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIfStatement.cs]: Note that not having to deal with different kind of initializers (simple variable/param/member initializer, array initializer, object initializer, collection initializer, member initializer, etc) does simplify flow analysis significantly, and lowering them into simple assignments followed by a declaration is definitely a value add.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIfStatement.cs
Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIfStatement.cs]: The IVariableDeclaration and the initializer aren't guaranteed to be in the same place, so clients will have to handle this anyway. If the client needs to report a diagnostic on the symbol's original definition, they'd have to go through `DeclaringSyntaxReferences` anyway.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIfStatement.cs
Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIfStatement.cs]: My plan this PR was to only handle `IVariableDeclarationGroup`, `IVariableDeclaration`, and `IVariableDeclarator`. If reviewing it is easier, I can also add in `IVariableInitializerOperation`, but I figured that would be handled in the same PR as the other initializer operations.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIfStatement.cs
Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIfStatement.cs]: Yes, but I am not sure what is the value add or necessity of removing `IVariableDeclaration` operation for flow analysis?

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIfStatement.cs
Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIfStatement.cs]: If lowering `IVariableInitializerOperation` is coming in next PR, that is fine. However, I am still not convinced why get rid of the declaration operations. What is the motivating factor behind this change?

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIfStatement.cs
Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIfStatement.cs]: There's no benefit to leaving it in, and it adds unnecessary nodes to the graph. We've already said that we're going to have variable lifetime information in a separate data structure, and all the `IVariableDeclaration` does (or really `IVariableDeclarator`, which is what I think you mean here) is provide that information.

Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIfStatement.cs]: @jinujoseph, we'll bring this up in the meeting today.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >declarator.Initializer [](start = 40, length = 22)

We should skip through IVariableInitializerOperation to the Value here

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >var [](start = 21, length = 3)

Please no ```var```s. The type must be spelled out unless it is spelled out explicitly in the initializer, which is not present here.

**Comment by AlekseyTs** [comment]
>             return new VariableInitializer(Visit(operation.Value), semanticModel: null, operation.Syntax, operation.Type, operation.ConstantValue, operation.IsImplicit);

We probably should throw unreachable here.

---
Refers to: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:1265 in 43561df. [](commit_id = 43561df0b021508ede099c7dd319f22f8d27ee0e, deletion_comment = False)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >operation.Initializer [](start = 60, length = 21)

Same here, IVariableInitializerOperation should be dropped completely.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: I'm planning on doing that skip in a separate PR for all the initializer types. I don't think the skip though should be here, the visitor for IVariableInitializer should just remove itself from the tree and return visit of it's child.

---
In reply to: [166775123](https://github.com/dotnet/roslyn/pull/24692#discussion_r166775123) [](ancestors = 166775123)

**Comment by 333fred** [comment]
>             return new VariableInitializer(Visit(operation.Value), semanticModel: null, operation.Syntax, operation.Type, operation.ConstantValue, operation.IsImplicit);

I'd rather have this and all the other initializer types just return `Visit(Operation.Value)`, less assumptions that need to be made.

---
In reply to: [363931562](https://github.com/dotnet/roslyn/pull/24692#issuecomment-363931562) [](ancestors = 363931562)

---
Refers to: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:1265 in 43561df. [](commit_id = 43561df0b021508ede099c7dd319f22f8d27ee0e, deletion_comment = False)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: It should be skipped here and in context of this PR. IVariableInitializerOperation doesn't represent a value and doesn't belong on the right hand side of the assignment. The remaining initializer types can wait as we haven't even decided what shape should the graph have for them and do not have an API for requesting it yet.

---
In reply to: [166777002](https://github.com/dotnet/roslyn/pull/24692#discussion_r166777002) [](ancestors = 166777002,166775123)

**Comment by AlekseyTs** [comment]
>             return new VariableInitializer(Visit(operation.Value), semanticModel: null, operation.Syntax, operation.Type, operation.ConstantValue, operation.IsImplicit);

> I'd rather have this and all the other initializer types just return Visit(Operation.Value), less assumptions that need to be made.

I don't think this is the right approach. This node doesn't represent a value and we should not be changing this during regular Visit.

---
In reply to: [363933264](https://github.com/dotnet/roslyn/pull/24692#issuecomment-363933264) [](ancestors = 363933264,363931562)

---
Refers to: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:1265 in 43561df. [](commit_id = 43561df0b021508ede099c7dd319f22f8d27ee0e, deletion_comment = False)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >new ExpressionStatement [](start = 36, length = 23)

I think we can skip creation of an ExpressionStatement here. It is fine to use simple assignment as a statement.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >public override IOperation VisitVariableDeclaration(IVariableDeclarationOperation operation, int? captureIdForResult) [](start = 8, length = 117)

```private void VisitVariableDeclaration(IVariableDeclarationOperation operation)```? And throw from ```public override IOperation VisitVariableDeclaration(IVariableDeclarationOperation operation, int? captureIdForResult)```

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: I considered this, but my thought was that for analyzer authors it would be helpful if:
```C#
int a = 1;
int b;
b = 1;
```
ended up with the same shape in the flow graph, namely an IExpressionStatementOperation wrapping an ISimpleAssignmentOperation.

---
In reply to: [166782118](https://github.com/dotnet/roslyn/pull/24692#discussion_r166782118) [](ancestors = 166782118)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >return null; [](start = 12, length = 12)

Is IVariableDeclarationGroupOperation always used as a statement? Consider asserting that it is equal to the ```_currentStatement```.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: With IOperation from syntax our hand was forced by syntax. There is no constraint like that for synthetic nodes, and I think we should keep amount synthesized nodes to a necessary minimum

---
In reply to: [166783230](https://github.com/dotnet/roslyn/pull/24692#discussion_r166783230) [](ancestors = 166783230,166782118)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: It's not used as the statement for `for` loops.

---
In reply to: [166783785](https://github.com/dotnet/roslyn/pull/24692#discussion_r166783785) [](ancestors = 166783785)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >VisitVariableDeclarationGroup [](start = 35, length = 29)

Please make necessary changes to ControlFlowGraphVerifier.CanBeInControlFlowGraph method to make sure that nodes that you've handled are in the right set. It is fine that there still can be scenarios that we do not handle, but as long as all current tests are passing, it would be good to accurately reflect final expectations in that function.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Will do.

---
In reply to: [166783214](https://github.com/dotnet/roslyn/pull/24692#discussion_r166783214) [](ancestors = 166783214)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIfStatement.cs
Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIfStatement.cs]: Per design discussion today, we'll fully remove the IVariableDeclarat(ion(Group))(or) from the resulting graph.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Updated.

---
In reply to: [166806899](https://github.com/dotnet/roslyn/pull/24692#discussion_r166806899) [](ancestors = 166806899)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Fixed.

---
In reply to: [166775498](https://github.com/dotnet/roslyn/pull/24692#discussion_r166775498) [](ancestors = 166775498)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: I'm going to leave this as is for now. During design discussion next week, we'll talk about removing all ExpressionStatements from the Control Flow graph, as they don't really serve a purpose.

---
In reply to: [166784473](https://github.com/dotnet/roslyn/pull/24692#discussion_r166784473) [](ancestors = 166784473,166783230,166782118)

**Comment by 333fred** [comment]
@mavasani @AlekseyTs updated to address feedback.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (mavasani) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Will we visit the `operation.Initializer` multiple times here for a declaration with multiple declarators sharing the same initializer? I feel we should visit them only once, and re-use the visited result for all assignments (probably move `operationInitializer` declaration outside the loop, and execute the visit only if `operationInitializer == null`?)

**Review by mavasani — APPROVED** [review]
LGTM, modulo the small issue of duplicate visits of the common initializer.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: > It's not used as the statement for for loops.

It well be. We are not handling loops properly at the moment. Please add an assert.

---
In reply to: [166784997](https://github.com/dotnet/roslyn/pull/24692#discussion_r166784997) [](ancestors = 166784997,166783785)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: We cannot reuse the node as it should have a different parent in each assignment

---
In reply to: [166980240](https://github.com/dotnet/roslyn/pull/24692#discussion_r166980240) [](ancestors = 166980240)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: To clarify, we cannot return null, unless we are visiting this node in a statement mode.

---
In reply to: [167016477](https://github.com/dotnet/roslyn/pull/24692#discussion_r167016477) [](ancestors = 167016477,166784997,166783785)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >captureIdForResult [](start = 70, length = 18)

We should not be propagating ```captureIdForResult``` here. We are not producing a value here, the parameter should be removed from the calls and the enclosing function.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Also, Value can contain control flow, which will add statements and basic blocks, simply reusing the returned node will not add these "side-effects".

---
In reply to: [167017796](https://github.com/dotnet/roslyn/pull/24692#discussion_r167017796) [](ancestors = 167017796,166980240)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >prototype [](start = 23, length = 9)

"prototype" should be in all caps

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >// prototype(dataflow): try/catch can get directly here, as can foreach control variables [](start = 12, length = 89)

Please remove this comment. try/catch won't get here once we handle it properly. No adjustments will  be necessary here.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >prototype [](start = 15, length = 9)

All caps for "prototype" please.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: @333fred Please add a PROTOTYPE comment that we should add a test with a control flow in a shared initializer. We can add it right now, but it I think it won't produce expected result because we don't handle control flow in object creation yet. 

---
In reply to: [167020856](https://github.com/dotnet/roslyn/pull/24692#discussion_r167020856) [](ancestors = 167020856,167017796,166980240)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IVariableDeclaration.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IVariableDeclaration.cs]: >int d = a ? b : c [](start = 8, length = 17)

It would be good to have a test with two locals initialized in the same statement and control flow in the second initializer.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IVariableDeclaration.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IVariableDeclaration.cs]: >IExpressionStatementOperation [](start = 8, length = 29)

Please add a PROTOTYPE comment, we need to discuss at design meeting whether we want to add assignments to constant locals at all. The fact that the symbol itself returns constant value could be sufficient.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IVariableDeclaration.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IVariableDeclaration.cs]: >prototype [](start = 11, length = 9)

All caps?

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Removed.

---
In reply to: [167019822](https://github.com/dotnet/roslyn/pull/24692#discussion_r167019822) [](ancestors = 167019822)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Added.

---
In reply to: [167018923](https://github.com/dotnet/roslyn/pull/24692#discussion_r167018923) [](ancestors = 167018923,167016477,166784997,166783785)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Fixed all instances of this.

---
In reply to: [167021108](https://github.com/dotnet/roslyn/pull/24692#discussion_r167021108) [](ancestors = 167021108)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Ok, removed.

---
In reply to: [167021700](https://github.com/dotnet/roslyn/pull/24692#discussion_r167021700) [](ancestors = 167021700)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Added.

---
In reply to: [167023541](https://github.com/dotnet/roslyn/pull/24692#discussion_r167023541) [](ancestors = 167023541,167020856,167017796,166980240)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IVariableDeclaration.cs
Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IVariableDeclaration.cs]: Added. FYI @jinujoseph for the design question.

---
In reply to: [167026093](https://github.com/dotnet/roslyn/pull/24692#discussion_r167026093) [](ancestors = 167026093)

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IVariableDeclaration.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IVariableDeclaration.vb]: >ISimpleAssignmentOperation [](start = 12, length = 26)

Please add a PROTOTYPE comment to confirm whether this node is good enough to indicate that this initialization of a static local happens only once, across all invocations of containing method

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IVariableDeclaration.cs
Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IVariableDeclaration.cs]: Added for C# and VB.

---
In reply to: [167024883](https://github.com/dotnet/roslyn/pull/24692#discussion_r167024883) [](ancestors = 167024883)

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IVariableDeclaration.vb
Comment (333fred) [reply src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IVariableDeclaration.vb]: Added. FYI @jinujoseph for the design question.

---
In reply to: [167030284](https://github.com/dotnet/roslyn/pull/24692#discussion_r167030284) [](ancestors = 167030284)

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 3).

**Comment by 333fred** [comment]
@AlekseyTs addressed feedback.

---
In reply to: [364210264](https://github.com/dotnet/roslyn/pull/24692#issuecomment-364210264) [](ancestors = 364210264)

**Review by AlekseyTs — APPROVED** [review]
LGTM (iteration 4)

**Comment by 333fred** [comment]
Ping @dotnet/roslyn-compiler for a second review.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IVariableDeclaration.cs
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IVariableDeclaration.cs]: Consider testing a variable assigned before it is declared.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (cston) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Consider splitting the comment across multiple lines. Same for other long lines in this method.

**Review by cston — APPROVED** [review]

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: The "advantages" of the new monitors.

---
In reply to: [167056137](https://github.com/dotnet/roslyn/pull/24692#discussion_r167056137) [](ancestors = 167056137)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IVariableDeclaration.cs
Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IVariableDeclaration.cs]: Added.

---
In reply to: [167055452](https://github.com/dotnet/roslyn/pull/24692#discussion_r167055452) [](ancestors = 167055452)

**Pull request merged**
